Windows multicore-related fixes

dra27 opened the pull request:
A few things found getting multicore-ocaml working on Windows:

 - `caml_stat_strdup_to_utf16` should use `caml_stat_alloc_noexc` not `malloc`. This would have caused problems when using the unloadable runtime on Windows, but is not a problem in backwards-compatible mode.
 - The debug runtime tags contain a long-standing overflow in 64-bit mode on Windows.
 - As a consequence of this, the debug runtime is now actually built on AppVeyor.
 - `caml_set_fields` incorrectly assumes `sizeof(long) == sizeof(intnat)`, as did the `magic` field of pool blocks. Both these only affect the debug runtime.

We already run the testsuite twice on AppVeyor, so I'm seeing what happens if the mingw32 testsuite uses the debug runtime and the msvc64 testsuite continues to use the normal runtime.

@shindere - does Inria CI test the debug runtime on Windows at all (I think this may have been mentioned recently, but I can't remember what the answer was)
@murmour - I'd be grateful if the unloadable runtime-related changes could have a quick glance, but I don't think I've damaged anything!
@damiendoligez - I wouldn't say that any of this is critical for 4.07.0

dra27 committed:
Incorrect malloc in caml_stat_strdup_to_utf16

Should use caml_stat_alloc_noexc instead.

dra27 committed:
Build and test debug runtime on AppVeyor

dra27 committed:
Fix building debug runtime on Windows

\r problems generating opnames.h and jumptbl.h

dra27 committed:
intnat and long are not the same thing

Various places - affecting debug runtime online - where long and intnat
have been assumed to be the same type.

shindere commented:
David Allsopp (2018/04/12 14:22 +0000):
> @shindere - does Inria CI test the debug runtime on Windows at all (I
> think this may have been mentioned recently, but I can't remember what
> the answer was)

Yet it's tested, I just double-checked to make sure.
See tools/ci-build:213:

    sed -i 's%RUNTIMED=.\+%RUNTIMED=true%' config/Makefile

Unrelated: would it make sense to create a "ci" directory, either at the
toplevel of the source tree, or under tools, and to put all the
ci-related scripts there? We could have one subdirectory of this
directory by continuous integration solution we are using and allthe
relevant file would be there, except of course thouse which need to be
at the root of the repository.

Any thoughts?

dra27 commented:
@shindere - yes, tidying the CI scripts up out of the root (leaving just .travis.yml and appveyor.yml) would be a good improvement!

dra27 commented:
I'm not certain why `__FILE__` is including the full path here (it should be just the C filename, as far as I know) and should not therefore be triggering these warnings. I'll have to look into that in more detail.

dra27 committed:
Fix display of Unicode paths in CAMLassert

dra27 commented:
I've looked at this further - the problem is the inline functions defined in `minor_gc.h` and `weak.h` - those functions will always have `__FILE__` as a full path. 

@damiendoligez: as a side-note, that isn't particularly great. I experimented with an alternate version where instead of using `__FILE__` in the headers, you specified the name - see https://github.com/dra27/ocaml/commit/4f5fb1388c11ca3f15cb0512c96e07aabd16ef82. A downside is that this is not recursive (so `Is_young`'s assertion would show up as "caml/weak.h", rather than "caml/address_class.h". It's not major, but it seems bad to have it that ocamlrund embeds the full path of where it was compiled.

Back to the plot here, through a series of twisted macro definitions, you can get `__FILE__` as a utf16 string and so that can then be passed `caml_stat_strdup_of_os` so I've essentially upgraded `caml_failed_assert` (@nojb - I think correctly?)

nojb commented:
> Back to the plot here, through a series of twisted macro definitions, you can get __FILE__ as a utf16 string and so that can then be passed caml_stat_strdup_of_os so I've essentially upgraded caml_failed_assert (@nojb - I think correctly?)

LGTM

dra27 committed:
Define CRT abort correctly for VS 2005-2010

Microsoft broke the definition of abort in stdlib.h in Visual Studio
2005, finally fixing it in Visual Studio 2012.

dra27 commented:
@xavierleroy is invited to roll his eyes at https://github.com/ocaml/ocaml/pull/1721/commits/cd0c3aca41b8139250b7506fcc0bd50744f74972. More seriously, I've checked my CRT archives, and that "hack" is consistent. The other option would be to use the same `#if` to guard a `return 0;` in `caml_failed_assert` or equally to disable warning C4716 around `caml_failed_assert`.

xavierleroy commented on byterun/caml/misc.h:
```diff
@@ -94,13 +94,33 @@ extern caml_timing_hook caml_major_slice_begin_hook, caml_major_slice_end_hook;
+   seem to have been superseded by a more sensible mechanism! */
+#define WIDEN2(x) L##x
+#define WIDEN(x) WIDEN2(x)
+#define __OSFILE__ WIDEN(__FILE__)
```
Could we use macro names less "obvious" than `WIDEN` and `WIDEN2`?  All OCaml-C stub files include `<caml/misc.h>`, directly or indirectly, hence I'm worried about macro name space pollution.  `CAML_WIDEN_STRING_LITERAL{,2}` perhaps?

dra27 replied:
Sorry for the lag - good suggestion, which I've followed.

xavierleroy commented on byterun/misc.c:
```diff
@@ -15,6 +15,16 @@
+#define _CRT_TERMINATE_DEFINED
+__declspec(noreturn) void __cdecl abort(void);
+#endif
+
```
Eyes rolled.

xavierleroy reviewed (commented):
Looks good to me, with one suggestion below.

damiendoligez commented:
@dra27 IIUC you're saying that **any** C source that uses `__FILE__` will fail to have reproducible builds. It looks like a PR for the maintainers of the C compilers. I don't think we should make our lives harder by adding an argument to CAMLassert.

dra27 committed:
Missing quotes in stdlib Makefile

dra27 commented:
@damiendoligez - that's not exactly practical, though, is it!! However, it's not C source files which have this problem - it's C **headers**. It's the inline functions defined (as in the C notion of defined, not declared) in `minor_gc.h` and `weak.h`. The patch in https://github.com/dra27/ocaml/commit/4f5fb1388c11ca3f15cb0512c96e07aabd16ef82 only uses this extra-argument version of `CAMLassert` in two places.

dra27 commented:
I'll merge this once the CI's passed (I've just been stung by the lack of https://github.com/ocaml/ocaml/pull/1721/commits/1221877f87749d975d43c7199bcf77046609964e) - I can open a separate GPR for the `CAMLnamedassert` thing, as that's not included in this GPR.

murmour commented:
> @murmour - I'd be grateful if the unloadable runtime-related changes could have a quick glance, but I don't think I've damaged anything!

I see no issues.

The change that replaces `malloc` with `caml_stat_alloc_noexc` fixes a bug that would cause a crash in the unloading-enabled runtime. Good catch!

dra27 commented:
@damiendoligez - can I cherry-pick ad4a004 to 4.07?

damiendoligez commented:
@dra27 go ahead. This will probably be the last thing to go into 4.07 before the release.

dra27 commented:
Done - https://github.com/ocaml/ocaml/commit/9e7419a6b9d75cc38213f1f168619aeede2a7bb2
